crates/ol/mempool/src/state.rs

```diff
@@ -0,0 +1,1933 @@
+    ///
+    /// This method:
+    /// 1. Revalidates existing transactions and removes invalid ones
+    /// 2. Verifies sequential seq_no ordering per account and removes transactions with gaps
+    /// 3. Finds all blocks that were rolled back (between old_tip and new_tip)
+    /// 4. Extracts transactions from those rolled-back blocks
+    /// 5. Converts them to mempool transactions (removing accumulator proofs)
+    /// 6. Adds them back to the mempool (they will be validated during add)
+    /// 7. Recalculates pending_seq_no tracking
+    async fn handle_reorg(
+        &mut self,
+        ctx: &MempoolContext,
+        old_tip: OLBlockCommitment,
+        new_tip: OLBlockCommitment,
+        state_accessor: &OLState,
+    ) -> OLMempoolResult<usize> {
+        // Step 1: Revalidate existing transactions and remove invalid ones
+        let invalid_txids = self.revalidate_all_transactions(ctx, state_accessor);
+        let mut removed_count = invalid_txids.len();
+        if !invalid_txids.is_empty() {
+            self.remove_transactions(ctx, &invalid_txids)?;
+        }
+
+        // Step 2: Verify sequential seq_no ordering and remove transactions with gaps
+        // For each account, collect SnarkAccountUpdate transactions and verify sequential ordering
+        let mut txs_by_account: HashMap<AccountId, Vec<(u64, OLTxId)>> = HashMap::new();
+        for (txid, entry) in &self.entries {
+            if let Some(base_update) = entry.tx.base_update() {
+                let target_account = entry.tx.target();
+                let seq_no = base_update.operation().seq_no();
+                txs_by_account
+                    .entry(target_account)
+                    .or_default()
+                    .push((seq_no, *txid));
+            }
+        }
+
+        // For each account, sort by seq_no and remove transactions with gaps
+        let mut gap_txids = Vec::new();
+        for (account, mut txs) in txs_by_account {
+            // Sort by seq_no
+            txs.sort_by_key(|(seq_no, _)| *seq_no);
+
+            // Get on-chain seq_no for this account
+            let on_chain_seq_no = state_accessor
+                .get_account_state(account)
+                .map_err(|e| {
+                    OLMempoolError::AccountStateAccess(format!(
+                        "Failed to get account state during reorg: {e}"
+                    ))
+                })?
+                .and_then(|account_state| {
+                    // Try to get as snark account, return 0 if not snark or error
+                    account_state
+                        .as_snark_account()
+                        .map(|snark_state| *snark_state.seqno().inner())
+                        .ok()
+                })
+                .unwrap_or(0);
+
+            // Verify sequential ordering starting from on_chain_seq_no
+            let mut expected_seq_no = on_chain_seq_no;
+            for (seq_no, txid) in txs {
```

Comment:
I feel this is too much work, we already no there are no gap for the txes that are _within_ the mempool. So you only need to check if the first tx, for each account, has a gap.